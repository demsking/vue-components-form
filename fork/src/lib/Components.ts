import { Dictionary, FieldKind, Component } from '@/types';

export type ComponentsType = 'form' | FieldKind;

export class Components {
  $: Dictionary<Component>;

  constructor () {
    this.$ = {
      form: 'form',
      boolean: 'input',
      integer: 'input',
      null: 'input',
      number: 'input',
      string: 'input',
      default: 'input'
    };
  }

  set(kind: ComponentsType, component: Component) {
    this.$[kind] = component;
  }

  get(kind: ComponentsType) {
    return this.$[kind] || this.$.default;
  }
}
