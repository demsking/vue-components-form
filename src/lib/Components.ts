import { Dict, Component, ComponentsType, ComponentsDeclaration } from '@/types';

export class Components implements ComponentsDeclaration {
  readonly $: Dict<Component>;

  constructor () {
    this.$ = {
      form: 'form',
      message: 'div',
      default: 'input'
    };
  }

  set(kind: ComponentsType, component: Component) {
    this.$[kind] = component;
  }

  get(kind: ComponentsType) {
    return this.$[kind] || this.$.default;
  }
}
