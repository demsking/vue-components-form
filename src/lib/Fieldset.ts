import { CreateElement } from 'vue';
import { IObjectChildDescriptor, ObjectField, ObjectFieldChild } from '@/types';

export const Fieldset = {
  renderChildren(h: CreateElement, field: ObjectField, children: ObjectFieldChild[] = field.childrenList) {
    return children
      .map((childField) => ({
        childField,
        childDescriptor: field.descriptor.getChildDescriptor(childField),
      }))
      .filter(({ childDescriptor }) => childDescriptor.visible)
      .map(({ childDescriptor, childField }) => h(childDescriptor.component, {
        key: childField.key,
        attrs: childDescriptor.attrs,
        props: {
          field: childField
        }
      }));
  },

  renderGroups(h: CreateElement, field: ObjectField) {
    return field.descriptor.childrenGroups.map(({ id, label, children }) => {
      const childrenNodes = Fieldset.renderChildren(h, field, children);
      const nodes = [
        h('div', {
          attrs: {
            'data-fs-group-nodes': childrenNodes.length
          }
        }, childrenNodes)
      ];

      if (label) {
        nodes.unshift(h('div', {
          attrs: {
            'data-fs-group-label': id
          }
        }, label));
      }

      return h('div', {
        attrs: {
          'data-fs-group': id
        }
      }, nodes);
    });
  }
};
