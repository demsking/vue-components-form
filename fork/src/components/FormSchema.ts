import { FormSchemaComponent } from '@/types';
import { UniqueId } from '@/lib/UniqueId';
import { Parser } from '@/parsers/Parser';
import { JsonSchema } from '@/types/jsonschema';
import { Objects } from '@/lib/Objects';
import { Components } from '@/lib/Components';
import { NativeElements } from "@/lib/NativeElements";
import { Descriptor } from "@/descriptors/Descriptor";

export const GLOBAL = {
  NativeElements,
  Descriptor
};

const FormSchema: FormSchemaComponent = {
  name: 'FormSchema',
  model: {
    prop: 'value',
    event: 'input'
  },
  props: {
    /**
     * The input JSON Schema object.
     */
    schema: { type: Object, required: true },

    /**
     * Use this directive to create two-way data bindings with the
     * component. It automatically picks the correct way to update the
     * element based on the input type.
     */
    value: {
      type: [ Number, String, Array, Object, Boolean ],
      default: undefined
    },

    /**
     * The id property of the Element interface represents the form's identifier,
     * reflecting the id global attribute.
     */
    id: {
      type: String,
      default: UniqueId.get('form')
    },

    /**
     * The name of the form. It must be unique among the forms in a document.
     */
    name: {
      type: String,
      default: undefined
    },

    /**
     * Use this prop to enable `search` landmark role to identify a section
     * of the page used to search the page, site, or collection of sites.
     */
    search: { type: Boolean, default: false },

    /**
     * Use this prop to overwrite the default Native HTML Elements with
     * custom components.
     */
    components: {
      type: Components,
      default: () => GLOBAL.NativeElements
    },

    /**
     * UI Schema Descriptor to use for rendering.
     *
     * @type {ScalarDescriptor | ObjectDescriptor | ArrayDescriptor | DescriptorConstructor}
     */
    descriptor: {
      type: [ Object, Function ],
      default: () => GLOBAL.Descriptor
    }
  },
  data: () => ({
    ref: UniqueId.get('form-schema'),
    field: null,
    default: {},
    data: undefined
  }),
  watch: {
    schema(value: JsonSchema) {
      this.load(value, this.value, false);
    }
  },
  created() {
    this.load(this.schema, this.value, false);
  },
  render(createElement) {
    if (this.field === null) {
      return null as any; // nothing to render
    }

    console.log(this.field)

    const root = createElement(this.field.component, {
      attrs: this.field.attrs.input,
      props: {
        field: this.field,
        value: this.data,
        components: this.components
      },
      on: {
        // [INPUT_ADDED_EVENT]: () => {
        //   this.$forceUpdate()
        // },
        input: (event: Event) => {
          // this.processFieldEvent(event);
          this.emitInputEvent();
        },
        change: (event: Event) => {
          // this.processFieldEvent(event);

          if (!Objects.equals(this.data, this.default)) {
            /**
             * Fired when a change to the element's value is committed
             * by the user.
             */
            this.$emit('change', this.data);
          }
        }
      }
    });

    const nodes = [ root ];

    if (this.$slots.default) {
      nodes.push(...this.$slots.default);
    }

    this.$nextTick(() => this.$emit('ready'));

    return createElement(this.components.get('form'), {
      ref: this.ref,
      attrs: {
        id: this.id,
        name: this.name,
        role: this.search ? 'search' : undefined,
        'aria-label': this.schema.title
      },
      props: {
        value: this.data,
        components: this.components
      },
      on: {
        reset: this.reset,
        submit: this.submit,
        invalid: this.invalid
      }
    }, nodes);
  },
  methods: {
    /**
     * Load the given `schema` with initial filled `value`.
     * Use this to load async schema.
     *
     * @param {object} schema - The JSON Schema object to load
     * @param {Number|String|Array|Object|Boolean} model - The initial data for the schema.
     *
     * @note `model` is not a two-way data bindings.
     * To get the form data, use the `v-model` directive.
     *
     * @note The default value of `model` is the initial model defined with the
     * `v-model` directive.
     */
    load (schema: JsonSchema, model: any = undefined, reset = true) {
      if (Objects.isEmpty(schema)) {
        this.field = null;
      } else {
        const descriptorConstructor = this.descriptor instanceof Function
          ? this.descriptor
          : GLOBAL.Descriptor.get;

        const descriptor = this.descriptor instanceof Function
          ? this.descriptor(schema)
          : GLOBAL.Descriptor.get(schema);

        const parser = Parser.get({
          schema: this.schema,
          model: typeof model === 'undefined' ? this.value : model,
          name: this.name,
          descriptor: descriptor,
          descriptorConstructor: descriptorConstructor
        });

        this.field = parser.field;
      }

      this.emitInputEvent();

      if (reset) {
        this.$nextTick(() => this.reset());
      }
    },

    /**
     * @private
     */
    emitInputEvent () {
      /**
       * Fired synchronously when the value of an element is changed.
       */
      this.$emit('input', this.data);
    },

    /**
     * Get the HTML form reference.
     *
     * @returns {HTMLFormElement|undefined} - Returns the HTML form element or `undefined` for empty object
     */
    form () {
      return this.$refs[this.ref];
    },

    /**
     * Returns true if the element's child controls satisfy their
     * validation constraints. When false is returned, cancelable invalid
     * events are fired for each invalid child and validation problems
     * are reported to the user.
     */
    reportValidity () {
      const controls: HTMLInputElement[] = this.form().elements as any;
      let validity = true;

      for (let i = 0; i < controls.length; i++) {
        if ('checkValidity' in controls[i]) {
          validity = validity && controls[i].checkValidity();
        }
      }

      return validity;
    },

    /**
     * @private
     */
    invalid (e) {
      /**
       * Fired when a submittable element has been checked and doesn't
       * satisfy its constraints. The validity of submittable elements is
       * checked before submitting their owner form, or after the
       * `checkValidity()` of the element or its owner form is called.
       */
      this.$emit('invalid', e);
    },

    /**
     * Reset the value of all elements of the parent form.
     */
    reset () {
      const form = this.form()

      if (form && 'reset' in form) {
        form.reset();
      }
    },

    /**
     * Send the content of the form to the server.
     * @private
     */
    submit (event) {
      if (this.reportValidity()) {
        /**
         * Fired when a form is submitted
         */
        this.$emit('submit', event);
      } else {
        this.invalid();
      }
    }
  }
};

export default FormSchema;
