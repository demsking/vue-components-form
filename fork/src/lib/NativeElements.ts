import { Components } from '@/lib/Components';
import { InputElement } from '@/components/InputElement';
import { FieldsetElement } from '@/components/FieldsetElement';
import { SelectElement } from '@/components/SelectElement';
import { TextareaElement } from '@/components/TextareaElement';

export const NativeElements = new Components();

NativeElements.set('boolean', InputElement);
NativeElements.set('string', InputElement);
NativeElements.set('radio', InputElement);
NativeElements.set('enum', FieldsetElement);
NativeElements.set('number', InputElement);
NativeElements.set('integer', InputElement);
NativeElements.set('object', FieldsetElement);
NativeElements.set('list', SelectElement);
NativeElements.set('textarea', TextareaElement);
