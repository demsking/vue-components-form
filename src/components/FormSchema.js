import { genId, loadFields } from '@/lib/parser'
import { equals, merge } from '@/lib/object'
import { Components as Instance, argName, inputName } from '@/lib/components'
import FormSchemaField from './FormSchemaField'

export const Components = Instance

export default {
  name: 'FormSchema',
  props: {
    /**
     * The JSON Schema object.
     */
    schema: { type: Object, default: () => ({}) },

    /**
     * Use this directive to create two-way data bindings with the component. It automatically picks the correct way to update the element based on the input type.
     * @model
     * @default {}
     */
    value: { type: Object, default: () => ({}) },

    /**
     * The URI of a program that processes the form information.
     */
    action: { type: String },

    /**
     * This property indicates whether the value of the control can be automatically completed by the browser. Possible values are: `off` and `on`.
     */
    autocomplete: { type: String },

    /**
     * When the value of the method attribute is post, enctype is the MIME type of content that is used to submit the form to the server. Possible values are:
     * - application/x-www-form-urlencoded: The default value if the attribute is not specified.
     * - multipart/form-data: The value used for an <input> element with the type attribute set to "file".
     * - text/plain (HTML5)
     */
    enctype: { type: String, default: 'application/x-www-form-urlencoded' },

    /**
     * The HTTP method that the browser uses to submit the form. Possible values are:
     * - post: Corresponds to the HTTP POST method ; form data are included in the body of the form and sent to the server.
     * - get: Corresponds to the HTTP GET method; form data are appended to the action attribute URI with a '?' as separator, and the resulting URI is sent to the server. Use this method when the form has no side-effects and contains only ASCII characters.
     */
    method: { type: String, default: 'post' },

    /**
     * This Boolean attribute indicates that the form is not to be validated when submitted.
     */
    novalidate: { type: Boolean },

    /**
     * Use this prop to overwrite the default Native HTML Elements for custom components.
     */
    components: {
      type: Components,
      default: () => new Components()
    }
  },
  data: () => ({
    ref: genId('form-schema'),
    schemaLoaded: { schema: {}, fields: [] },
    default: {},
    error: null,
    data: {},
    inputValues: {}
  }),
  created () {
    this.loadSchema(this.schema)
  },
  render (createElement) {
    const { schema, fields } = this.schemaLoaded
    const components = this.components
    const nodes = []

    if (schema.title) {
      nodes.push(createElement(components.$.title.component, schema.title))
    }

    if (schema.description) {
      nodes.push(createElement(
        components.$.description.component, schema.description))
    }

    if (this.error) {
      nodes.push(createElement(components.$.error.component, this.error))
    }

    const formNodes = fields.map((field) => {
      const value = this.data[field.attrs.name]

      return createElement(FormSchemaField, {
        props: { field, value, components },
        on: {
          input: (event) => {
            const target = event.target
            const data = event.target.value
            const eventInput = { field, data, target }

            if (field.isArrayField) {
              this.onInputArrayValue(eventInput)
            } else {
              this.onInput(eventInput)
            }
          },
          change: (event) => {
            const target = event.target
            const data = event.target.value
            const eventInput = { field, data, target }

            if (field.isArrayField) {
              this.onInputArrayValue(eventInput, false)
            } else {
              this.onInput(eventInput, false)
            }

            this.changed()
          }
        }
      })
    })

    if (formNodes.length) {
      if (this.$slots.default) {
        formNodes.push(createElement(
          components.$.buttonswrapper.component, this.$slots.default))
      }

      nodes.push(createElement(components.$.form.component, {
        ref: this.ref,
        [argName(components.$.form)]: {
          action: this.action,
          enctype: this.enctype,
          method: this.method,
          autocomplete: this.autocomplete,
          novalidate: this.novalidate
        },
        on: {
          reset: this.reset,
          submit: this.submit,
          invalid: this.invalid
        }
      }, formNodes))
    }

    return createElement(components.$.formwrapper.component, nodes)
  },
  methods: {
    /**
     * Load the given JSON Schema. Use this to update the initial schema.
     */
    loadSchema (schema) {
      const fields = []

      loadFields(schema, fields)
      this.loadDefaultValues(fields)

      this.schemaLoaded = { schema, fields }
    },

    /**
     * @private
     */
    loadDefaultValues (fields) {
      this.default = {}
      this.inputValues = {}

      fields.forEach((field) => {
        const { type, name } = field.attrs

        this.default[name] = field.schemaType === 'boolean'
          ? typeof this.value[name] === 'boolean'
            ? this.value[name]
            : field.attrs.checked === true
          : this.value[name] || field.attrs.value

        if (field.isArrayField) {
          if (!Array.isArray(this.default[name])) {
            this.default[name] = []
          } else {
            this.default[name] = this.default[name].filter((value, i) => {
              this.inputValues[inputName(field, i)] = value
              return value !== undefined
            })
          }

          field.itemsNum = type === 'checkbox'
            ? field.items.length
            : field.minItems
        }
      })

      const data = {}

      merge(data, this.default)

      this.data = data

      /**
       * Fired synchronously when the value of an element is changed.
       */
      this.$emit('input', this.data)
    },

    /**
     * @private
     */
    onInput (event, triggerInputEvent = true) {
      if (event.field.schemaType === 'boolean') {
        event.data = event.target.checked
      }

      this.data[event.field.attrs.name] = event.data

      if (triggerInputEvent) {
        /**
         * Fired synchronously when the value of an element is changed.
         */
        this.$emit('input', this.data)
      }
    },

    /**
     * @private
     */
    onInputArrayValue (event, triggerInputEvent = true) {
      if (event.field.attrs.type === 'checkbox') {
        if (event.target.checked) {
          if (!this.data[event.field.attrs.name].includes(event.data)) {
            this.data[event.field.attrs.name].push(event.data)
          }
        } else {
          const index = this.data[event.field.attrs.name].indexOf(event.data)

          if (index > -1) {
            this.data[event.field.attrs.name].splice(index, 1)
          }
        }
      } else {
        const index = event.target.getAttribute('data-fs-index')
        const key = inputName(event.field, index)

        this.inputValues[key] = event.data

        const values = []

        for (let i = 0; i < event.field.itemsNum; i++) {
          const currentValue = this.inputValues[inputName(event.field, i)]

          if (currentValue) {
            values.push(currentValue)
          }
        }

        this.data[event.field.attrs.name] = values
      }

      if (triggerInputEvent) {
        /**
         * Fired synchronously when the value of an element is changed.
         */
        this.$emit('input', this.data)
      }
    },

    /**
     * @private
     */
    changed () {
      if (!equals(this.data, this.default)) {
        /**
         * Fired when a change to the element's value is committed by the user.
         */
        this.$emit('change', this.data)
      }
    },

    /**
     * Get the HTML form reference.
     */
    form () {
      return this.$refs[this.ref]
    },

    /**
     * Returns true if the element's child controls satisfy their validation constraints. When false is returned, cancelable invalid events are fired for each invalid child and validation problems are reported to the user.
     */
    reportValidity () {
      const controls = this.form().elements
      let validity = true

      for (let i = 0; i < controls.length; i++) {
        if ('checkValidity' in controls[i]) {
          validity = validity && controls[i].checkValidity()
        }
      }

      return validity
    },

    /**
     * Checks whether the form has any constraints and whether it satisfies them. If the form fails its constraints, the browser fires a cancelable `invalid` event at the element, and then returns false.
     * @aliasof reportValidity
     */
    checkValidity () {
      return this.reportValidity()
    },

    /**
     * @private
     */
    invalid (e) {
      /**
       * Fired when a submittable element has been checked and doesn't satisfy its constraints. The validity of submittable elements is checked before submitting their owner form, or after the `checkValidity()` of the element or its owner form is called.
       */
      this.$emit('invalid', e)
    },

    /**
     * Reset the value of all elements of the parent form.
     */
    reset () {
      for (let key in this.inputValues) {
        delete this.inputValues[key]
      }

      this.schemaLoaded.fields.forEach((field) => {
        const { name } = field.attrs

        this.$set(this.data, name, this.default[name])

        if (field.isArrayField) {
          this.data[name].forEach((value, i) => {
            this.inputValues[inputName(field, i)] = value
          })
        }
      })
    },

    /**
     * Send the content of the form to the server.
     */
    submit (event) {
      if (this.checkValidity()) {
        /**
         * Fired when a form is submitted
         */
        this.$emit('submit', event)
      }
    },

    /**
     * Set a message error.
     */
    setErrorMessage (message) {
      this.error = message
    },

    /**
     * clear the message error.
     */
    clearErrorMessage () {
      this.error = null
    }
  }
}
