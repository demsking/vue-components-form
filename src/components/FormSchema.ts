import { VNode } from 'vue';
import { FormSchemaComponent } from '@/types';
import { UniqueId as UniqueIdLib } from '@/lib/UniqueId';
import { Objects as ObjectsLib } from '@/lib/Objects';
import { Components as ComponentsLib } from '@/lib/Components';
import { Parser as ParserLib } from '@/parsers/Parser';
import { NativeElements } from '@/lib/NativeElements';
import { NativeDescriptor } from '@/lib/NativeDescriptor';

import '@/parsers';

export const GLOBAL = {
  Elements: NativeElements,
  Descriptor: NativeDescriptor
};

export const Objects = ObjectsLib;
export const UniqueId = UniqueIdLib;
export const Components = ComponentsLib;
export const Parser = ParserLib;

const FormSchema: FormSchemaComponent = {
  name: 'FormSchema',
  model: {
    prop: 'value',
    event: 'input'
  },
  props: {
    /**
     * The input JSON Schema object.
     */
    schema: { type: Object, required: true },

    /**
     * Use this directive to create two-way data bindings with the
     * component. It automatically picks the correct way to update the
     * element based on the input type.
     */
    value: {
      type: [ Number, String, Array, Object, Boolean ],
      default: undefined
    },

    /**
     * The id property of the Element interface represents the form's identifier,
     * reflecting the id global attribute.
     */
    id: {
      type: String,
      default: UniqueId.get('form')
    },

    /**
     * The name of the form. It must be unique among the forms in a document.
     */
    name: {
      type: String,
      default: undefined
    },

    /**
     * When set to true (default), checkbox inputs and nested object inputs will
     * automatically include brackets at the end of their names
     * (e.g. name="Multicheckbox-Value1[]".
     * Setting this property to false, disables this behaviour.
     */
    bracketedObjectInputName: {
      type: Boolean,
      default: true
    },

    /**
     * Use this prop to enable `search` landmark role to identify a section
     * of the page used to search the page, site, or collection of sites.
     */
    search: { type: Boolean, default: false },

    /**
     * Indicates whether the form elements are disabled or not.
     */
    disabled: { type: Boolean, default: false },

    /**
     * Use this prop to overwrite the default Native HTML Elements with
     * custom components.
     */
    components: {
      type: Components,
      default: () => GLOBAL.Elements
    },

    /**
     * UI Schema Descriptor to use for rendering.
     *
     * @type {ScalarDescriptor|ObjectDescriptor|ArrayDescriptor|DescriptorConstructor}
     */
    descriptor: {
      type: [ Object, Function ],
      default: () => GLOBAL.Descriptor.get
    }
  },
  data: () => ({
    key: undefined,
    ref: UniqueId.get('formschema'),
    initialModel: undefined,
    ready: false
  }),
  computed: {
    fieldId() {
      return `${this.id}-field`;
    },
    descriptorConstructor() {
      return typeof this.descriptor === 'function'
        ? this.descriptor
        : GLOBAL.Descriptor.get;
    },
    schemaDescriptor() {
      return typeof this.descriptor === 'function'
        ? this.descriptor(this.schema)
        : this.descriptor || GLOBAL.Descriptor.get(this.schema);
    },
    parser() {
      return Parser.get({
        schema: this.schema,
        model: this.initialModel,
        name: this.name,
        id: this.fieldId,
        required: true,
        descriptor: this.schemaDescriptor,
        descriptorConstructor: this.descriptorConstructor,
        bracketedObjectInputName: this.bracketedArrayInputName,
        onChange: this.emitInputEvent,
        requestRender: this.update
      });
    },
    field() {
      return this.parser instanceof Parser ? this.parser.field : null;
    },
    listeners(): Record<string, Function | Function[]> {
      const on: any = { ...this.$listeners };

      on.reset = on.reset ? [ on.reset ] : [];

      on.reset.unshift(() => {
        if (this.field) {
          this.field.input.reset();
        }
      });

      // remove the injected vue's input event
      // to prevent vue errors on the submit event
      delete on.input;

      return on;
    }
  },
  watch: {
    schema: {
      handler() {
        this.ready = false;
        this.initialModel = this.clone(this.value);
      },
      immediate: true
    }
  },
  render(createElement) {
    if (this.field === null || this.ready === false) {
      return null as any; // nothing to render
    }

    this.field.input.attrs.disabled = this.disabled;

    const props = {
      field: this.field,
      disabled: this.disabled
    };

    const key = this.key || this.field.key;
    const element = createElement(this.field.input.component, { key, props });
    const nodes = [ element ];

    if (this.$slots.default) {
      nodes.push(...this.$slots.default);
    }

    return createElement(this.components.get('form'), {
      ref: this.ref,
      attrs: {
        id: this.id,
        name: this.name,
        role: this.search ? 'search' : undefined
      },
      props: props,
      on: this.listeners
    }, nodes);
  },
  methods: {
    clone(value): unknown {
      if (Objects.isObject(value)) {
        const object = value instanceof Array ? [] : {};

        return Objects.assign(object, value as any);
      }

      return value;
    },

    /**
     * Get the HTML form reference.
     *
     * @returns {HTMLFormElement|VNode|undefined} - Returns the HTML form element or `undefined` for empty object
     */
    form(): HTMLFormElement | VNode | undefined {
      return this.$refs[this.ref] as any;
    },

    /**
     * @private
     */
    emitInputEvent(value: unknown) {
      /**
       * Fired synchronously when the value of an element is changed.
       */
      this.$emit('input', value);

      this.$nextTick(() => {
        this.ready = true;
      });
    },

    /**
     * @private
     */
    update([ field ]) {
      this.key = field.key;
    }
  }
};

export default FormSchema;
