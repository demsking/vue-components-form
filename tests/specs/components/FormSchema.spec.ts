/* eslint-disable @typescript-eslint/camelcase */

import Vue from 'vue';

import { mount } from '@vue/test-utils';

import FormSchema, {
  GLOBAL,
  UniqueId,
  Objects,
  Components,
  Fieldset,
  Parser,
  UIDescriptor,
  NativeComponents
} from '@/components/FormSchema';

import { JsonSchema } from '@/types/jsonschema';

import { UniqueId as UniqueIdLib } from '@/lib/UniqueId';
import { Objects as ObjectsLib } from '@/lib/Objects';
import { Components as ComponentsLib } from '@/lib/Components';
import { NativeComponents as NativeComponentsLib } from '@/lib/NativeComponents';
import { Fieldset as FieldsetLib } from '@/lib/Fieldset';
import { Parser as ParserLib } from '@/parsers/Parser';
import { UIDescriptor as UIDescriptorLib } from '@/descriptors/UIDescriptor';

import { NativeElements } from '@/lib/NativeElements';

const DEFAULT_ID = 'id-form';

function getWrapper(options: any): any {
  const {
    id = DEFAULT_ID, schema, value, children, ...props
  }: any = options;

  return mount({
    functional: true,
    render(h: Function) {
      return h(FormSchema, {
        props: {
          id, value, schema, ...props
        }
      }, children);
    }
  });
}

const schema: JsonSchema = {
  type: 'string',
  title: 'String',
  description: 'A String'
};

describe('components/FormSchema', () => {
  describe('exported libraries', () => {
    it('should successfully export the GLOBAL.Fieldset library', () => {
      expect(GLOBAL.Elements).toBe(NativeElements);
    });

    it('should successfully export the UniqueId library', () => {
      expect(UniqueId).toBe(UniqueIdLib);
    });

    it('should successfully export the Objects library', () => {
      expect(Objects).toBe(ObjectsLib);
    });

    it('should successfully export the Components library', () => {
      expect(Components).toBe(ComponentsLib);
    });

    it('should successfully export the NativeComponents library', () => {
      expect(NativeComponents).toBe(NativeComponentsLib);
    });

    it('should successfully export the Fieldset library', () => {
      expect(Fieldset).toBe(FieldsetLib);
    });

    it('should successfully export the Parser library', () => {
      expect(Parser).toBe(ParserLib);
    });

    it('should successfully export the UIDescriptor library', () => {
      expect(UIDescriptor).toBe(UIDescriptorLib);
    });
  });

  describe('props/computed', () => {
    it('should have setted schema', (done) => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      Vue.nextTick(() => {
        expect(vm.schema).toBe(schema);
        done();
      });
    });

    it('vm.id should be defined', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.id).toBeDefined();
    });

    it('vm.name should be defined', () => {
      const name = 'string';
      const wrapper = getWrapper({ schema, name });
      const { vm } = wrapper;

      expect(vm.name).toBeDefined();
    });

    it('vm.bracketedObjectInputName should be defined', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.bracketedObjectInputName).toBeTruthy();
    });

    it('vm.search should be defined', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.search).toBeFalsy();
    });

    it('vm.disabled should be defined', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.disabled).toBeFalsy();
    });

    it('vm.components should be defined', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.components).toBeDefined();
    });

    it('vm.descriptor should be an empty object', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.descriptor).toEqual({});
    });

    it('vm.fieldId should be defined', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.fieldId).toBeDefined();
    });
  });

  describe('should successfully render component', () => {
    it('as an undefined VNode when before the nextTick', () => {
      const wrapper = getWrapper({ schema });
      const { vm } = wrapper;

      expect(vm.ready).toBeFalsy();
      expect(wrapper.html()).toBeUndefined();
    });

    it('after the nextTick', () => {
      const wrapper = getWrapper({ schema });
      const expected = '<form id="id-form"><div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div><p id="id-form-field-helper" data-fs-helper="true">A String</p></div></div></form>';

      return Vue.nextTick().then(() => {
        const { vm } = wrapper;

        expect(vm.ready).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot(expected);
      });
    });

    it('should emit an input event with an undefined model', () => {
      const wrapper = getWrapper({ schema });

      expect(Object.keys(wrapper.emitted())).toEqual([ 'input' ]);
      expect(wrapper.emitted().input).toEqual([ [ undefined ] ]);
    });

    it('should emit an input event with a defined model', () => {
      const value = 'gohan';
      const wrapper = getWrapper({ schema, value });

      expect(Object.keys(wrapper.emitted())).toEqual([ 'input' ]);
      expect(wrapper.emitted().input).toEqual([ [ value ] ]);
    });

    it('as an undefined VNode with an undefined schema.type', () => {
      const wrapper = getWrapper({ schema: { ...schema, type: undefined } });

      expect(wrapper.html()).toBeUndefined();
    });

    it('with a default slot', () => {
      const children = 'checking...';
      const wrapper = getWrapper({ schema, children });
      const expected = '<form id="id-form"><div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div><p id="id-form-field-helper" data-fs-helper="true">A String</p></div></div>checking...</form>';

      return Vue.nextTick().then(() => expect(wrapper.html()).toMatchSnapshot(expected));
    });

    it('with only a scalar schema', () => {
      const wrapper = getWrapper({ schema });
      const expected = '<form id="id-form"><div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div><p id="id-form-field-helper" data-fs-helper="true">A String</p></div></div></form>';

      return Vue.nextTick().then(() => expect(wrapper.html()).toMatchSnapshot(expected));
    });

    it('with a non scalar schema', (done) => {
      const objectSchema = {
        type: 'object',
        properties: {
          string: schema
        }
      };

      const wrapper = getWrapper({ schema: objectSchema });
      const expected = { id: 'id-form' };

      Vue.nextTick(() => {
        expect(wrapper.attributes()).toEqual(expected);
        done();
      });
    });

    it('with { schema, search }', (done) => {
      const search = true;
      const wrapper = getWrapper({ schema, search });
      const expected = { id: 'id-form', role: 'search' };

      Vue.nextTick(() => {
        expect(wrapper.attributes()).toEqual(expected);
        done();
      });
    });

    it('with { schema, disabled }', (done) => {
      const objectSchema = {
        type: 'object',
        properties: {
          string: schema
        }
      };

      const wrapper = getWrapper({ schema: objectSchema, disabled: true });
      const expected = '<form id="id-form"><fieldset id="id-form-field" disabled="disabled"><div data-fs-kind="string" data-fs-type="text" data-fs-field="string"><label id="id-form-field-string-label" for="id-form-field-string">String</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-form-field-string" type="text" name="string" aria-labelledby="id-form-field-string-label" aria-describedby="id-form-field-string-helper"></div><span id="id-form-field-string-helper" data-fs-helper="true">A String</span></div></div></fieldset></form>';

      Vue.nextTick(() => {
        expect(wrapper.html()).toMatchSnapshot(expected);
        done();
      });
    });
  });

  it('vm.form() should be return a defined VNode', (done) => {
    const wrapper = getWrapper({ schema });

    Vue.nextTick(() => {
      const { vm } = wrapper.find('form');

      expect(vm.form()).toBeDefined();
      done();
    });
  });

  it('vm.update(updatedFields) should be called', () => {
    const schema: JsonSchema = {
      type: 'object',
      properties: {
        credit_card: {
          type: 'number'
        },
        billing_address: {
          type: 'string'
        }
      },
      dependencies: {
        credit_card: [
          'billing_address'
        ]
      }
    };

    const wrapper = getWrapper({ schema });
    const { vm } = wrapper;
    const previousKey = vm.key;

    vm.update = jest.fn();
    vm.parser.field.children.credit_card.setValue(123);

    expect(vm.key).not.toBe(previousKey);
  });
});
