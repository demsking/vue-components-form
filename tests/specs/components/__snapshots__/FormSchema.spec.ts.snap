// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FormSchema should successfully render component with a default slot: <form id="id-form"><div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div><p id="id-form-field-helper">A String</p></div></div>checking...</form> 1`] = `
<form id="id-form">
  <div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label>
    <div data-fs-wrapper="2">
      <div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div>
      <p id="id-form-field-helper">A String</p>
    </div>
  </div>checking...
</form>
`;

exports[`components/FormSchema should successfully render component with only a scalar schema: <form id="id-form"><div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label><div data-fs-wrapper="2"><div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div><p id="id-form-field-helper">A String</p></div></div></form> 1`] = `
<form id="id-form">
  <div data-fs-kind="string" data-fs-type="text" data-fs-required="true"><label id="id-form-field-label" for="id-form-field">String</label>
    <div data-fs-wrapper="2">
      <div data-fs-input="text"><input id="id-form-field" type="text" required="required" aria-required="true" aria-labelledby="id-form-field-label" aria-describedby="id-form-field-helper"></div>
      <p id="id-form-field-helper">A String</p>
    </div>
  </div>
</form>
`;
